backend/x86_peephole/x86_peephole_utils.ml

```diff
@@ -0,0 +1,398 @@
+    | Asm_targets.Asm_directives.Directive.Bytes _
+    | Asm_targets.Asm_directives.Directive.Cfi_adjust_cfa_offset _
+    | Asm_targets.Asm_directives.Directive.Cfi_def_cfa_offset _
+    | Asm_targets.Asm_directives.Directive.Cfi_endproc
+    | Asm_targets.Asm_directives.Directive.Cfi_offset _
+    | Asm_targets.Asm_directives.Directive.Cfi_startproc
+    | Asm_targets.Asm_directives.Directive.Cfi_remember_state
+    | Asm_targets.Asm_directives.Directive.Cfi_restore_state
+    | Asm_targets.Asm_directives.Directive.Cfi_def_cfa_register _
+    | Asm_targets.Asm_directives.Directive.Comment _
+    | Asm_targets.Asm_directives.Directive.Const _
+    | Asm_targets.Asm_directives.Directive.Direct_assignment _
+    | Asm_targets.Asm_directives.Directive.File _
+    | Asm_targets.Asm_directives.Directive.Global _
+    | Asm_targets.Asm_directives.Directive.Indirect_symbol _
+    | Asm_targets.Asm_directives.Directive.Loc _
+    | Asm_targets.Asm_directives.Directive.New_line
+    | Asm_targets.Asm_directives.Directive.Private_extern _
+    | Asm_targets.Asm_directives.Directive.Size _
+    | Asm_targets.Asm_directives.Directive.Sleb128 _
+    | Asm_targets.Asm_directives.Directive.Space _
+    | Asm_targets.Asm_directives.Directive.Type _
+    | Asm_targets.Asm_directives.Directive.Uleb128 _
+    | Asm_targets.Asm_directives.Directive.Protected _
+    | Asm_targets.Asm_directives.Directive.Hidden _
+    | Asm_targets.Asm_directives.Directive.Weak _
+    | Asm_targets.Asm_directives.Directive.External _
+    | Asm_targets.Asm_directives.Directive.Reloc _ ->
+      false)
+  | Ins _ -> false
+
+let get_cells cell n =
+  let rec loop acc remaining current_opt =
+    if remaining <= 0
+    then List.rev acc
+    else
+      match current_opt with
+      | None -> List.rev acc
+      | Some current -> loop (current :: acc) (remaining - 1) (DLL.next current)
+  in
+  loop [] n (Some cell)
+
+let equal_args arg1 arg2 =
+  match arg1, arg2 with
+  | Imm i1, Imm i2 -> Int64.equal i1 i2
+  | Sym s1, Sym s2 -> String.equal s1 s2
+  | Reg8L r1, Reg8L r2 -> equal_reg64 r1 r2
+  | Reg8H r1, Reg8H r2 -> equal_reg8h r1 r2
+  | Reg16 r1, Reg16 r2 -> equal_reg64 r1 r2
+  | Reg32 r1, Reg32 r2 -> equal_reg64 r1 r2
+  | Reg64 r1, Reg64 r2 -> equal_reg64 r1 r2
+  | Regf rf1, Regf rf2 -> equal_regf rf1 rf2
+  | Mem addr1, Mem addr2 -> equal_addr addr1 addr2
+  | Mem64_RIP (t1, s1, i1), Mem64_RIP (t2, s2, i2) ->
+    equal_data_type t1 t2 && String.equal s1 s2 && i1 = i2
+  | ( ( Imm _ | Sym _ | Reg8L _ | Reg8H _ | Reg16 _ | Reg32 _ | Reg64 _ | Regf _
+      | Mem _ | Mem64_RIP _ ),
+      _ ) ->
+    false
+
+let is_register = function
+  | Reg8L _ | Reg8H _ | Reg16 _ | Reg32 _ | Reg64 _ | Regf _ -> true
+  | Imm _ | Sym _ | Mem _ | Mem64_RIP _ -> false
+
+let is_safe_self_move_arg = function
+  | Reg8L _ | Reg8H _ | Reg16 _ | Reg64 _ -> true
+  | Imm _ | Sym _ | Reg32 _ | Regf _ | Mem _ | Mem64_RIP _ -> false
+
+let is_safe_for_dead_register_opt = function
```

Comment:
could use a more discriptive name - I think the condition is that writing to it never effects aliases?